### PR TITLE
 #162725264 users should favorite articles

### PR DESCRIPTION
**Pull request by armstrongsouljah**
#### What does this PR do?
This pr adds functionality to enable authenticated users to favorite unfavorite articles. 

#### Description of the task to be implemented.
Articles that tickle our fancy are worth favoriting and those deemed distasteful are worth unfavoriting.

#### How should this be manually tested?
- Fetch this branch to your local machine by running `git fetch origin ft-favorite-articles-162725264`  on your terminal
- Check into this branch by running `git checkout ft-favorite-articles-162725264` 
- Run `python manage.py migrate` to introduce a new field
- Run your server by running `python manage.py runserver`
- Favorite an article by making a `POST` request on any article `/api/articles/article-slug/favorite`
screenshot 2019-01-25 at 20 01 08

- View favorite  articles by making a `GET` request on any article `/api/articles/favorites`
screenshot 2019-01-28 at 11 07 24

- Unfavorite an article by making a `DELETE` request on any article `/api/articles/article-slug/favorite`
screenshot 2019-01-25 at 20 04 09

#### Related pivotal tracker stories.
[#162725264](https://www.pivotaltracker.com/story/show/162725264)

#### Postman Collections
[Run in Postman](https://app.getpostman.com/run-collection/9720cdb1ee29cf146477)

### CODE REVIEW

File: authors/apps/utils/messages.py [hunk 1]
```diff
@@ -5,3 +5,9 @@
     'permission_denied': "You do not have permission to perform this action",
     'delete_msg': "{} successfully deleted"
 }
+
+
+favorite_actions_messages = {
+    'favorited':'Article has been added to favorites',
+    'un_favorited': 'Article has been removed from favorites'
+}
```
Comment (malfahad) [inline authors/apps/utils/messages.py:24]: These can be joined into a single message with help of the `.format()` `string` method

**Review by malfahad — DISMISSED** [review]

File: authors/apps/utils/messages.py (hunk 1 above)
Comment (malfahad) [reply authors/apps/utils/messages.py:24]: But I guess it can be done during the code cleanup session

Comment (armstrongsouljah) [reply authors/apps/utils/messages.py:24]: I get your point but in this case, I would want to render a different message based on the action performed. Let me see if I can refactor to meet your request.

**Review by MuhweziDeo — DISMISSED** [review]

File: authors/apps/articles/serializers.py [hunk 2]
```diff
@@ -86,3 +87,15 @@ def get_slug(self, obj):
...
+
+
+class FavoriteSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Article
+        fields = [
+            'title',
+            'body',
+            
+        ]
```
Comment (malfahad) [inline authors/apps/articles/serializers.py:137]: Consider adding a slug to the returned fields. This will allow a user to link back to the main article through their favorites.

**Review by malfahad — CHANGES_REQUESTED** [review]

File: authors/apps/articles/serializers.py (hunk 2 above)
Comment (armstrongsouljah) [reply authors/apps/articles/serializers.py:137]: Okay, let me do that. Thanks.

**Review by malfahad — DISMISSED** [review]

**Comment by ja-odur** [comment]
@armstrongsouljah please update your documentation

**Comment by armstrongsouljah** [comment]
> @armstrongsouljah please update your documentation

@ja-odur Postman docs?

**Comment by armstrongsouljah** [comment]
> > @armstrongsouljah please update your documentation
> 
> @ja-odur Postman docs?

let me resolve that.

**Comment by ja-odur** [comment]
screen shot 2019-01-25 at 18 44 57

@armstrongsouljah  only return the `usernames`instead of `ids` in the favorites
also include an endpoint for the user to get their favorite articles.

File: authors/apps/articles/models.py [hunk 3]
```diff
@@ -1,8 +1,39 @@
...
-from authors.apps.authentication.models import User
+from authors.apps.utils.messages import error_messages, favorite_actions_messages
+
+
+class ArticleManager(models.Manager):
+    def handle_favorite_actions(self, request_user_obj, article_slug):
+        """ 
+        Enales both favoriting and unfavoriting of articles 
+        Args:
+            request_obj:
```
Comment (ja-odur) [inline authors/apps/articles/models.py]: @armstrongsouljah Please fully define all arguments

File: authors/apps/profiles/serializers.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 from .models import Profile
 from authors.apps.authentication.models import User
 
+
```
Comment (ja-odur) [inline authors/apps/profiles/serializers.py:6]: @armstrongsouljah unnecessary change here

File: authors/apps/utils/messages.py [hunk 5]
```diff
@@ -14,3 +14,11 @@
     'follow message': 'You have followed {}', 
     'unfollow message': 'You have unfollowed {}'
 }
+
+
+favorite_actions_messages = {
```
Comment (ja-odur) [inline authors/apps/utils/messages.py:19]: @armstrongsouljah Nice

File: authors/apps/articles/models.py (hunk 3 above)
Comment (armstrongsouljah) [reply authors/apps/articles/models.py]: okay. Let me work on that.

**Comment by mariamiah** [comment]
screen shot 2019-01-25 at 7 14 55 pm
 @armstrongsouljah please add an option for checking out into the branch too between these two phrases

**Comment by mariamiah** [comment]
screen shot 2019-01-25 at 7 20 21 pm
Please adjust the first and last line and please add some screenshots too

**Comment by d-rita** [comment]
include check out branch in your PR.

**Comment by armstrongsouljah** [comment]
> screen shot 2019-01-25 at 7 14 55 pm
> 
> @armstrongsouljah please add an option for checking out into the branch too between these two phrases

Done

**Comment by armstrongsouljah** [comment]
> include check out branch in your PR.

already done that.

**Comment by armstrongsouljah** [comment]
> screen shot 2019-01-25 at 7 20 21 pm
> 
> Please adjust the first and last line and please add some screenshots too

will do.

**Comment by mariamiah** [comment]
Tested the end points and they work perfectly, good job

**Comment by armstrongsouljah** [comment]
> screen shot 2019-01-25 at 18 44 57
> 
> @armstrongsouljah only return the `usernames`instead of `ids` in the favorites
> also include an endpoint for the user to get their favorite articles.

this has been resolved.

File: authors/apps/profiles/serializers.py (hunk 4 above)
Comment (armstrongsouljah) [reply authors/apps/profiles/serializers.py:6]: This happened due to a rebase that led to the loss of my code. Thus the change.

**Review by d-rita — DISMISSED** [review]

**Review by mariamiah — DISMISSED** [review]
LGTM

**Comment by ja-odur** [comment]
@armstrongsouljah the get favorites endpoint url is `{{base-url}}/api/favorites`, I think you should change it to `{{base-url}}/api/articles/favorites` so that the users are well aware of what they are getting.

File: authors/apps/articles/import_statements.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+from rest_framework import generics
```
Comment (ja-odur) [inline authors/apps/articles/import_statements.py]: @armstrongsouljah this file is unnecessary, please undo this change and leave the imports at the top of the files as they were.

File: authors/apps/articles/migrations/0001_initial.py [hunk 7]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (ja-odur) [inline authors/apps/articles/migrations/0001_initial.py:1]: @armstrongsouljah please undo this change

File: authors/apps/articles/renderers.py [hunk 8]
```diff
@@ -6,3 +6,6 @@ class ArticleJSONRenderer(AppJSONRenderer):
 
 class BookmarkJSONRenderer(AppJSONRenderer):
     name = 'bookmarks'
+
+class FavortiesJsonRenderer(AppJSONRenderer):
+    name = "favorites"
```
Comment (ja-odur) [inline authors/apps/articles/renderers.py]: @armstrongsouljah add new line

File: authors/apps/articles/serializers.py [hunk 9]
```diff
@@ -21,14 +21,32 @@ class ArticleSerializer(serializers.ModelSerializer):
     author = AuthorProfileSerializer(read_only=True)
     read_time = serializers.SerializerMethodField()
     share_links = serializers.SerializerMethodField()
-
+    favorites = serializers.SerializerMethodField()
```
Comment (ja-odur) [inline authors/apps/articles/serializers.py:24]: @armstrongsouljah please add this line back for readability purpose

File: authors/apps/articles/serializers.py [hunk 10]
```diff
@@ -21,14 +21,32 @@ class ArticleSerializer(serializers.ModelSerializer):
...
                   'body', 'image', 'author', 'read_time', 'share_links',
                   'likes_count', 'dislikes_count', 'average_rating')
 
+
+    def generate_usernames(self,profiles):
+        """
+        Resolve profile ids into usernames
+
+        Args:
+          profiles : Django queryset
```
Comment (ja-odur) [inline authors/apps/articles/serializers.py]: @armstrongsouljah Please follow the docstring convention shared

File: authors/apps/articles/views.py [hunk 11]
```diff
@@ -1,21 +1,4 @@
...
-from rest_framework.response import Response
-from authors.apps.utils.messages import error_messages
-from authors.apps.profiles.models import Profile
-from authors.apps.utils.custom_permissions.permissions import (
-    check_if_is_author
-)
-from .paginators import ArticleLimitOffSetPagination
-from .utils import get_like_status, get_usernames
-from rest_framework.exceptions import NotFound
+from .import_statements import *
```
Comment (ja-odur) [inline authors/apps/articles/views.py]: @armstrongsouljah undo the changes

Comment (armstrongsouljah) [reply authors/apps/articles/views.py]: Let me do that. However, we had opted to do this to avoid the 255 max limit. But let me put them back to how they were. Thank you.

File: authors/apps/articles/serializers.py (hunk 10 above)
Comment (armstrongsouljah) [reply authors/apps/articles/serializers.py]: Okay let me resolve these.

File: authors/apps/articles/import_statements.py (hunk 6 above)
Comment (armstrongsouljah) [reply authors/apps/articles/import_statements.py]: working on it.

File: authors/apps/articles/migrations/0001_initial.py (hunk 7 above)
Comment (armstrongsouljah) [reply authors/apps/articles/migrations/0001_initial.py:1]: This happened while I was resolving merge conflicts. I had rebased and lost the code for the current feature, thus the change above.

**Review by calvinpete — COMMENTED** [review]
@armstrongsouljah  please add the url for the GET favourite articles method in the postman collection

File: authors/apps/articles/serializers.py (hunk 10 above)
Comment (armstrongsouljah) [reply authors/apps/articles/serializers.py]: Done. @ja-odur

**Review by calvinpete — DISMISSED** [review]
@armstrongsouljah Works fine to me

File: authors/apps/articles/models.py [hunk 12]
```diff
@@ -3,8 +3,47 @@
...
+        Args:
+            request_user:  currently loggedin user
+            article_slug: Article model instance
+
+        """
+        article_slug = kwargs.get("article_slug")
+        request_user = kwargs.get("request_user")
+        user_ = Profile.objects.get(user=request_user)
+        article_to_unfavorite = self.model.objects.get(slug=article_slug)
+        if article_to_unfavorite.favorites.count() > 0:
```
Comment (ja-odur) [inline authors/apps/articles/models.py]: @armstrongsouljah Have tried using
```suggestion
        if article_to_unfavorite.favorites.count():
```

Comment (armstrongsouljah) [reply authors/apps/articles/models.py]: No i haven't. Let me try and see if it renders the same output, I will change it to that.

Comment (armstrongsouljah) [reply authors/apps/articles/models.py]: It actually works the same way. Let me push the changes.

**Review by ja-odur — APPROVED** [review]
LGTM

**Review by mariamiah — APPROVED** [review]
Looks Good To Me

**Review by calvinpete — APPROVED** [review]
LGTM
